automerge-frontend/src/state_tree/mod.rs

```diff
@@ -1,309 +1,173 @@
 
 /// Represents the result of running a local operation (i.e one that happens within the frontend
 /// before any interaction with a backend).
 pub(crate) struct LocalOperationResult {
-    /// The new state tree after the operation is executed
-    new_state: StateTree,
     /// Any operations which need to be sent to the backend to reconcile this change
     pub new_ops: Vec<amp::Op>,
 }
 
-impl LocalOperationResult {
-    pub(crate) fn new_state(&self) -> StateTree {
-        self.new_state.clone()
-    }
-}
-
 #[derive(Debug, Clone, PartialEq)]
 pub(crate) struct StateTree {
-    objects: im_rc::HashMap<amp::ObjectId, StateTreeComposite>,
+    root_props: HashMap<String, MultiValue>,
     cursors: Cursors,
 }
 
+impl Default for StateTree {
+    fn default() -> Self {
+        Self {
+            root_props: HashMap::new(),
+            cursors: Cursors::new(),
+        }
+    }
+}
+
 impl StateTree {
     pub fn new() -> StateTree {
         StateTree {
-            objects: im_rc::hashmap! {
-                amp::ObjectId::Root => StateTreeComposite::Map(StateTreeMap {
-                    object_id: amp::ObjectId::Root,
-                    props: im_rc::HashMap::new(),
-                })
-            },
+            root_props: HashMap::new(),
             cursors: Cursors::new(),
         }
     }
 
-    pub fn apply_root_diff(&self, diff: amp::RootDiff) -> Result<StateTree, error::InvalidPatch> {
-        match self.objects.get(&ObjectId::Root) {
-            Some(StateTreeComposite::Map(m)) => {
-                let diffapp = m.apply_diff(&DiffToApply {
-                    parent_key: &"",
-                    parent_object_id: &amp::ObjectId::Root,
-                    diff: &diff.props,
-                    current_objects: self.objects.clone(),
-                })?;
-                Ok(StateTree {
-                    objects: diffapp.change.objects().union(self.objects.clone()),
-                    cursors: self.cursors.clone(),
-                })
-            }
-            Some(o) => Err(error::InvalidPatch::MismatchingObjectType {
-                object_id: ObjectId::Root,
-                actual_type: Some(o.obj_type()),
-                patch_expected_type: Some(amp::ObjType::map()),
-            }),
-            None => {
-                let map = StateTreeMap {
-                    object_id: ObjectId::Root,
-                    props: im_rc::HashMap::new(),
-                };
-                let diffapp = map.apply_diff(&DiffToApply {
-                    parent_key: &"",
-                    parent_object_id: &amp::ObjectId::Root,
-                    diff: &diff.props,
-                    current_objects: self.objects.clone(),
-                })?;
-                Ok(StateTree {
-                    objects: diffapp.change.objects(),
-                    cursors: self.cursors.clone(),
-                })
+    pub fn check_diff(&self, diff: &amp::RootDiff) -> Result<(), error::InvalidPatch> {
```

Comment:
My thinking (and from the flamegraphs) was that if we don't check the patch before applying it we need to make a clone of the state to ensure that we don't corrupt it in the case of an erroneous patch. However, cloning (especially large documents) gets expensive so if we can cheaply check the diff then we can just apply it without risk of it being faulty (but this does make checking it a bit difficult).

Alternatively, we should just make a clone of the smallest sections that are changed before applying the patch but that still could get complicated.

Reply:
Check then apply makes sense in that context. I think we could reflect this in the type system to make the intent clearer. Maybe some kind of lightweight typestate pattern like this?:

```rust
struct Diff<T> {
    inner: automerge_protocol::Diff,
    phantom: PhantomData<T>
}

struct Unchecked;
struct Checked;

impl Diff<Unchecked> {
    fn check(...) -> Diff<Checked>
}

impl StateTree... {
    fn apply_diff(&mut self, Diff<Checked>) -> ...
}
```

This would ensure that the check then apply pipeline is followed and provide a few hooks to place  documentation about what the expected use of these methods is.